src/qibolab/instruments/zhinst.py

```diff
@@ -431,19 +434,34 @@ def register_readout_line(self, qubit, intermediate_frequency):
         self.signal_map[f"acquire{q}"] = self.device_setup.logical_signal_groups[f"q{q}"].logical_signals[
             "acquire_line"
         ]
-        self.calibration[f"/logical_signal_groups/q{q}/acquire_line"] = lo.SignalCalibration(
-            oscillator=lo.Oscillator(
-                frequency=intermediate_frequency,
-                modulation_type=lo.ModulationType.SOFTWARE,
-            ),
-            local_oscillator=lo.Oscillator(
-                uid="lo_shfqa_a" + str(q),
-                frequency=int(qubit.readout.local_oscillator.frequency),
-            ),
-            range=qubit.feedback.power_range,
-            port_delay=self.time_of_flight * NANO_TO_SECONDS,
-            threshold=qubit.threshold,
-        )
+
+        if qubit.kernel_path:
+            self.kernels[q] = qubit.kernel_path / f"kernels_q{q}.npz"
+        if self.kernels[q].is_file() and options.acquisition_type == AcquisitionType.DISCRIMINATION:
+            self.calibration[f"/logical_signal_groups/q{q}/acquire_line"] = lo.SignalCalibration(
+                oscillator=None,
+                range=qubit.feedback.power_range,
+                port_delay=self.time_of_flight * NANO_TO_SECONDS,
+            )
+        elif self.kernels[q].is_file() and not options.acquisition_type == AcquisitionType.DISCRIMINATION:
+            self.calibration[f"/logical_signal_groups/q{q}/acquire_line"] = lo.SignalCalibration(
+                oscillator=lo.Oscillator(
+                    frequency=intermediate_frequency,
+                    modulation_type=lo.ModulationType.SOFTWARE,
+                ),
+                range=qubit.feedback.power_range,
+                port_delay=self.time_of_flight * NANO_TO_SECONDS,
+                threshold=qubit.threshold,  # To keep compatibility with angle and threshold discrimination
+            )
+        elif not self.kernels[q].is_file() and not options.acquisition_type == AcquisitionType.DISCRIMINATION:
+            self.calibration[f"/logical_signal_groups/q{q}/acquire_line"] = lo.SignalCalibration(
+                oscillator=lo.Oscillator(
+                    frequency=intermediate_frequency,
+                    modulation_type=lo.ModulationType.SOFTWARE,
+                ),
+                range=qubit.feedback.power_range,
+                port_delay=self.time_of_flight * NANO_TO_SECONDS,
+            )
```

Comment:
The `if` at the moment is something like:
```py
if A and B:
    mydict["key"] = MyClass(...)
elif A and not B:
    mydict["key"] = MyClass(...)
elif not A and not B:
    mydict["key"] = MyClass(...)
```

There are a few improvements possible:
1. instead of repeating the conditions, you could make them hierarchical, or even assigning the result to a shorter variable
2. the dict, key, and class are always the same, the only things changing are the extra options, so you could just define those in the `if`, and then create the instance and set the item once only
3. the fourth case (`not A and B`) is not taken into account, and undocumented

Reply:
Ideally, this should be simplified if the calibration teams switches to kernels instead of threshold and angle and the middle if would be removed. I will try to make it nicer for the time being.